appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/OpenApiService.java

```diff
@@ -251,4 +239,68 @@ private static String getContextRoot(ApplicationInfo appInfo) {
                 .collect(toSet());
     }
 
+    private class OpenApiMapping {
+
+        private final ApplicationInfo appInfo;
+        private final OpenApiConfiguration appConfig;
+        private final CompletableFuture<OpenAPI> document;
+
+        private OpenApiMapping(ApplicationInfo appInfo) {
+            this.appInfo = appInfo;
+            this.appConfig = new OpenApiConfiguration(appInfo.getAppClassLoader());
+            this.document = new CompletableFuture<>();
+        }
+
+        private ApplicationInfo getAppInfo() {
+            return appInfo;
+        }
+
+        private OpenApiConfiguration getAppConfig() {
+            return appConfig;
+        }
+
+        private synchronized OpenAPI getDocument() throws InterruptedException, ExecutionException, TimeoutException {
+            if (!document.isDone()) {
+                executor.submit(new OpenApiBuilder(this, document));
```

Comment:
The method is synchronized for this reason, it can only be run in one thread at a time. Since the wait loop is in there, all threads will wait for the document to be built.

Reply:
I've just added the second check you asked for anyway, since there's no harm in adding it!

Reply:
I believe it's already thread safe. It submits the task to the executor, and then waits in the same method for it to be finished.

Reply:
It's not entirely thread-safe, see my comment below on line 283.